### PR TITLE
Add cuDNN sliding window and set_deterministic_algorithm

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR integrates cuDNN sliding window feature and cudnn-frontend `set_deterministic_algorithm` feature into TE.

Sliding window support:
```
   backend                 |      window_size       | diagonal alignment
---------------------------------------------------------------------------------
FlashAttention             | (-1, -1) or (>=0, >=0) | bottom right
FusedAttention             | (-1,  0) or (>=0, 0)   | top left 
UnfusedDotProductAttention | (-1, -1) or (>=0, >=0) | both;
                           |                        | converts window_size to an 'arbitrary' mask 
```

Determinism support:
```
backend                      | deterministic
---------------------------------------------
FlashAttention               |
    flash-attn >=2.0, <2.4.1 | no
    flash-attn >=2.4.1       | yes
FusedAttention               |
    sub-backend 0            | yes
    sub-backend 1            | workspace optimization path and sm90+: yes;
                             | otherwise: no
    sub-backend 2            | no
UnfusedDotProductAttention   | yes
```
For `FusedAttention` sub-backend 1, to enable determinism, please
- set `NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT=1` for cuDNN >=8.9.5 and <9.0.0
- set `NVTE_ALLOW_NONDETERMINISTIC_ALGO=0` for cuDNN >=9.0.0

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Integrate cuDNN sliding window and cudnn-frontend `set_deterministic_algorithm` into TE
- Add sliding window support to `UnfusedDotProductAttention`
- Cache `get_attention_backend` results so `DotProductAttention` only queries it once and reduces CPU overhead
- Add top-left-diagonal causal mask to `UnfusedDotProductAttention`
- Add top-left-diagonal ALiBi to `get_alibi`
- Improve control logic for causal masks, ALiBi, determinism, sliding window in `get_attention_backend` 

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by cyanguwa** [comment]
@Wong4j let me know what you think of the determinism logic. Thanks!

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -4875,35 +5139,29 @@ def __init__(
...
-        self.use_fused_attention = int(
-            os.getenv("NVTE_FUSED_ATTN", "1")
-        ) and self.device_compute_capability >= (8, 0)
-        if int(os.getenv("NVTE_FUSED_ATTN", "1")) == 0:
-            self.logger.debug("Disabling FusedAttention due to NVTE_FUSED_ATTN=0")
-        if self.device_compute_capability < (8, 0):
-            self.logger.debug("Disabling FusedAttention for compute capability < sm80")
+        # To use the workspace optimization path for determinism, please
+        # set NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT=1 for cuDNN >=8.9.5 and <9.0.0,
+        # and set NVTE_ALLOW_NONDETERMINISTIC_ALGO=1 for cuDNN >=9.0.0.
```
Comment (Wong4j) [inline transformer_engine/pytorch/attention.py:5148]: It should be `NVTE_ALLOW_NONDETERMINISTIC_ALGO=0` here.

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py [hunk 2]
```diff
@@ -1106,6 +1151,13 @@ def fused_attn_bwd(
                 type of the bias; {"no_bias", "pre_scale_bias", "post_scale_bias", "alibi"}
     attn_mask_type: str, default = "padding"
                 type of the attention mask; {"padding", "causal", "padding_causal", "no_mask"}
+    window_size: Tuple[int, int], default = (-1, -1)
+                sliding window size for local attention, where query at position i attends to keys
+                in [i + seqlen_k - seqlen_q - window_size[0], i + seqlen_k - seqlen_q
+                + window_size[1]] inclusive. Special cases (-1, -1) and (-1, 0) mean no sliding
```
Comment (zlsh80826) [inline transformer_engine/pytorch/cpp_extensions/fused_attn.py:1157]: Why there are two special cases for the same "no sliding". This seems hard to maintain.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 3]
```diff
@@ -208,6 +220,10 @@ void fused_attn_arbitrary_seqlen_fwd_impl(
                          .set_causal_mask_bottom_right(is_bottom_right)
                          .set_attn_scale(attn_scale);
 
+      if (cudnn_runtime_version >= 90200 && window_size_left != s_q) {
+        sdpa_options.set_sliding_window_length(window_size_left);
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:224]: I didn't find out where `window_size_right` is passed to cuDNN. Will cuDNN support that? If not, what is the behavior when user passed different `window_size_right` and `window_size_left`, e.g. (2, 3).

File: transformer_engine/common/include/transformer_engine/fused_attn.h [hunk 4]
```diff
@@ -248,6 +256,9 @@ void nvte_fused_attn_fwd_qkvpacked(const NVTETensor QKV, const NVTETensor Bias,
  *  \param[in]     qkv_layout               QKV tensor's layout.
  *  \param[in]     bias_type                Bias type.
  *  \param[in]     attn_mask_type           Attention mask type.
+ *  \param[in]     window_size_left         Sliding window size (the left half).
+ *  \param[in]     window_size_right        Sliding window size (the right half).
+ *  \param[in]     deterministic             Whether to execute with deterministic behaviours.
```
Comment (zlsh80826) [inline transformer_engine/common/include/transformer_engine/fused_attn.h:261]: typo: one more space
```suggestion
 *  \param[in]     deterministic            Whether to execute with deterministic behaviours.
```

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 3 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:224]: cuDNN attention only supports (>=0, 0) which is with causal mask implied. But I've requested to them that we want both left and right half of the window. I added `window_size_right` in the API to be future-proof, but right now any `window_size_right != 0` would disable the cuDNN path.

Comment (zlsh80826) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:224]: Make sense, but would be great to add NVTE_CHECK here before cuDNN support it.

Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:224]: We do have `nvte_get_fused_attn_backend` [here](https://github.com/cyanguwa/TransformerEngine/blob/28c883e1d05557d407fd4a9bb0a23537b0a60d62/transformer_engine/common/fused_attn/fused_attn.cpp#L175-L176) which is called by every `nvte_fused_attn_fwd/bwd_xxx` API. I think it's sufficient.

File: transformer_engine/common/include/transformer_engine/fused_attn.h (hunk 4 above)
Comment (cyanguwa) [reply transformer_engine/common/include/transformer_engine/fused_attn.h]: Done

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py (hunk 2 above)
Comment (cyanguwa) [reply transformer_engine/pytorch/cpp_extensions/fused_attn.py:1157]: The two special cases are no sliding window and causal mask, not just no sliding window. I added more words to describe them: 

Special cases (-1, -1) and (-1, 0) mean no sliding window and causal mask specifically. Both `causal` and 
`causal_bottom_right` masks map to `window_size = (-1, 0)` and Transformer Engine distinguishes them based on `attn_mask_type`. Similar to `attn_mask_type`, `window_size` can be overridden by `window_size` in `forward` as well.

**Review by zlsh80826 — APPROVED** [review]
LGTM

**Comment by zlsh80826** [comment]
That would be great if we document following somewhere
 - FA w/ deterministic: require more workspace and slower

**Pull request merged**